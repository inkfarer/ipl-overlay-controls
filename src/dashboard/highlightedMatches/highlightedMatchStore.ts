import { NodeCGBrowser } from 'nodecg/browser';
import { createStore, Store, useStore } from 'vuex';
import cloneDeep from 'lodash/cloneDeep';
import { InjectionKey } from 'vue';
import {
    HighlightedMatches,
    TournamentData
} from 'schemas';
import { SelectOptions } from '../types/select';
import { TournamentDataSource } from 'types/enums/tournamentDataSource';

const highlightedMatches = nodecg.Replicant<HighlightedMatches>('highlightedMatches');
const tournamentData = nodecg.Replicant<TournamentData>('tournamentData');

export const highlightedMatchReps = [ highlightedMatches, tournamentData ];

export interface HighlightedMatchStore {
    highlightedMatches: HighlightedMatches;
    tournamentData: TournamentData;
}

export const highlightedMatchStore = createStore<HighlightedMatchStore>({
    state: {
        highlightedMatches: null,
        tournamentData: null
    },
    mutations: {
        setState(store, { name, val }: { name: string, val: unknown }): void {
            this.state[name] = cloneDeep(val);
        }
    },
    actions: {
        async getHighlightedMatches(store, { options }: { options: SelectOptions }): Promise<void> {
            const getAllMatches = options.some(option => option.value === 'all');

            if (getAllMatches) {
                return nodecg.sendMessage('getHighlightedMatches', { getAllMatches });
            } else {
                const values = options.map(option => option.value);

                switch (store.state.tournamentData.meta.source) {
                    case TournamentDataSource.BATTLEFY:
                        return nodecg.sendMessage('getHighlightedMatches', {
                            getAllMatches: false,
                            stages: values
                        });
                    case TournamentDataSource.SMASHGG:
                        return nodecg.sendMessage('getHighlightedMatches', {
                            getAllMatches: false,
                            streamIDs: values.map(value => parseInt(value))
                        });
                    default:
                        throw new Error(`Cannot import data from source '${store.state.tournamentData.meta.source}'`);
                }
            }
        },
        async setNextMatch(store, { teamAId, teamBId }: { teamAId: string, teamBId: string }): Promise<void> {
            return nodecg.sendMessage('setNextRound', { teamAId, teamBId });
        }
    }
});

export const highlightedMatchStoreKey: InjectionKey<Store<HighlightedMatchStore>> = Symbol();

export function useHighlightedMatchStore(): Store<HighlightedMatchStore> {
    return useStore(highlightedMatchStoreKey);
}
