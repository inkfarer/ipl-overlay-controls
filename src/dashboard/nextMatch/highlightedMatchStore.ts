import { NodeCGBrowser } from 'nodecg/browser';
import { HighlightedMatches, TournamentData } from 'schemas';
import { SelectOptions } from '../types/select';
import { TournamentDataSource } from 'types/enums/tournamentDataSource';
import { SetNextRoundRequest } from 'types/messages/rounds';
import { defineStore } from 'pinia';

const highlightedMatches = nodecg.Replicant<HighlightedMatches>('highlightedMatches');
const tournamentData = nodecg.Replicant<TournamentData>('tournamentData');

export const highlightedMatchReps = [ highlightedMatches, tournamentData ];

export interface HighlightedMatchStore {
    highlightedMatches: HighlightedMatches;
    tournamentData: TournamentData;
}

export const useHighlightedMatchStore = defineStore('highlightedMatches', {
    state: () => ({
        highlightedMatches: null,
        tournamentData: null
    } as HighlightedMatchStore),
    actions: {
        async getHighlightedMatches({ options }: { options: SelectOptions }): Promise<void> {
            const getAllMatches = options.some(option => option.value === 'all');

            if (getAllMatches) {
                return nodecg.sendMessage('getHighlightedMatches', { getAllMatches });
            } else {
                const values = options.map(option => option.value);

                switch (this.tournamentData.meta.source) {
                    case TournamentDataSource.BATTLEFY:
                        return nodecg.sendMessage('getHighlightedMatches', {
                            getAllMatches: false,
                            stages: values
                        });
                    case TournamentDataSource.SMASHGG:
                        return nodecg.sendMessage('getHighlightedMatches', {
                            getAllMatches: false,
                            streamIDs: values.map(value => parseInt(value))
                        });
                    default:
                        throw new Error(`Cannot import data from source '${this.tournamentData.meta.source}'`);
                }
            }
        },
        async setNextMatch(data: SetNextRoundRequest): Promise<void> {
            return nodecg.sendMessage('setNextRound', data);
        }
    }
});
