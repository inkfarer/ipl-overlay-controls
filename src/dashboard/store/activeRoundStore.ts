import { ActiveRound, SwapColorsInternally } from 'schemas';
import { createStore, Store, useStore } from 'vuex';
import cloneDeep from 'lodash/cloneDeep';
import { InjectionKey } from 'vue';
import { GameWinner } from 'types/enums/gameWinner';
import { SetActiveColorRequest, SetWinnerRequest, SwapRoundColorRequest } from 'types/messages/activeRound';
import { SetRoundRequest } from 'types/messages/rounds';
import { ActiveRoundGame } from 'types/activeRoundGame';

const activeRound = nodecg.Replicant<ActiveRound>('activeRound');
const swapColorsInternally = nodecg.Replicant<SwapColorsInternally>('swapColorsInternally');

export const activeRoundReps = [ activeRound, swapColorsInternally ];

export interface ActiveRoundStore {
    activeRound: ActiveRound;
    swapColorsInternally: SwapColorsInternally;
}

export const activeRoundStore = createStore<ActiveRoundStore>({
    state: {
        activeRound: null,
        swapColorsInternally: null
    },
    mutations: {
        setState(store, { name, val }: { name: string, val: unknown }): void {
            this.state[name] = cloneDeep(val);
        }
    },
    actions: {
        setWinner(store, { winner }: { winner: GameWinner }): void {
            nodecg.sendMessage('setWinner', { winner });
        },
        removeWinner(): void {
            nodecg.sendMessage('removeWinner');
        },
        setActiveColor(store, request: SetActiveColorRequest): void {
            nodecg.sendMessage('setActiveColor', request);
        },
        swapColors(): void {
            swapColorsInternally.value = !swapColorsInternally.value;
        },
        setWinnerForIndex(store, { index, winner }: { index: number, winner: GameWinner }): void {
            nodecg.sendMessage('setWinner', { winner, roundIndex: index } as SetWinnerRequest);
        },
        resetActiveRound(): void {
            nodecg.sendMessage('resetActiveRound');
        },
        setActiveRound(store, request: SetRoundRequest): void {
            nodecg.sendMessage('setActiveRound', request);
        },
        swapRoundColor(store, request: SwapRoundColorRequest): void {
            nodecg.sendMessage('swapRoundColor', request);
        },
        updateActiveGames(store, games: ActiveRoundGame[]): void {
            nodecg.sendMessage('updateActiveGames', { games });
        },
        switchToNextColor(): void {
            nodecg.sendMessage('switchToNextColor');
        },
        switchToPreviousColor(): void {
            nodecg.sendMessage('switchToPreviousColor');
        },
        async getNextAndPreviousColors(): Promise<void> {
            return nodecg.sendMessage('getNextAndPreviousColors');
        }
    }
});

export const activeRoundStoreKey: InjectionKey<Store<ActiveRoundStore>> = Symbol();

export function useActiveRoundStore(): Store<ActiveRoundStore> {
    return useStore(activeRoundStoreKey);
}
