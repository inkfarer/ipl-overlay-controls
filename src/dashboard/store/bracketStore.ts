import { Bracket } from '@tourneyview/common';
import { defineStore } from 'pinia';

const bracketData = nodecg.Replicant<Bracket | undefined>('bracketData');

export const bracketReps = [ bracketData ];

export interface BracketStore {
    bracketData: Bracket | undefined;
}

export const useBracketStore = defineStore('brackets', {
    state: () => ({
        bracketData: undefined
    } as BracketStore),
    actions: {
        resetBracketData() {
            bracketData.value = undefined;
        }
    }
});
