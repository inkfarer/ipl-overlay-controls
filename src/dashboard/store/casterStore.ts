import { NodeCGBrowser } from 'nodecg/browser';
import { Caster, Casters, RadiaSettings } from 'schemas';
import { generateId } from '../../helpers/generateId';
import { defineStore } from 'pinia';

const casters = nodecg.Replicant<Casters>('casters');
const radiaSettings = nodecg.Replicant<RadiaSettings>('radiaSettings');

export const castersReps = [ casters, radiaSettings ];

export interface CasterStore {
    casters: Casters;
    uncommittedCasters: Casters;
    radiaSettings: RadiaSettings;
}

const defaultCaster = {
    name: 'New Caster',
    twitter: '',
    pronouns: '?/?'
};

export const useCasterStore = defineStore('casters', {
    state: () => ({
        casters: {},
        uncommittedCasters: {},
        radiaSettings: {
            enabled: null,
            guildID: null,
            updateOnImport: null
        }
    } as CasterStore),
    actions: {
        updateCaster({ id, newValue }: { id: string, newValue: Caster }): void {
            casters.value[id] = newValue;
        },
        removeUncommittedCaster(id: string): void {
            delete this.uncommittedCasters[id];
        },
        addUncommittedCaster({ id, caster }: { id: string, caster?: Caster }): void {
            this.uncommittedCasters[id] = caster ?? defaultCaster;
        },
        addDefaultCaster(): string {
            const id = generateId();
            this.addUncommittedCaster({ id });
            return id;
        },
        async saveUncommittedCaster({ id, caster }: { id: string, caster: Caster }): Promise<string> {
            const newId = await nodecg.sendMessage('saveCaster', caster);
            this.removeUncommittedCaster(id);
            return newId;
        },
        async loadCastersFromVc(): Promise<void> {
            const result = await nodecg.sendMessage('getLiveCommentators');
            if (result.extra && result.extra.length > 0) {
                for (let i = 0; i < result.extra.length; i++) {
                    const extraCaster = result.extra[i];
                    const id = extraCaster.discord_user_id;
                    delete extraCaster.discord_user_id;
                    this.addUncommittedCaster({ id, caster: extraCaster });
                }
            }
        },
        showCasters() {
            nodecg.sendMessage('mainShowCasters');
        },
        async removeCaster(id: string): Promise<void> {
            return nodecg.sendMessage('removeCaster', { id });
        }
    }
});
