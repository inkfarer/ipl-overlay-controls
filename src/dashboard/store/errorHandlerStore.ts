import { App } from 'vue';
import { generateId } from '../../helpers/generateId';
import { defineStore } from 'pinia';

export const useErrorHandlerStore = defineStore('errorHandler', {
    state: () => ({
        recentErrors: {}
    }),
    actions: {
        removeRecentError({ key }: { key: string }): void {
            delete this.recentErrors[key];
        },
        handleError({ err, info }: { err: unknown, info: string }): void {
            console.error(`Got error from '${info}': \n`, err);

            if (Object.keys(this.recentErrors).length >= 2) return;

            const id = generateId();
            this.recentErrors[id] = err;
            window.setTimeout(() => {
                delete this.recentErrors[id];
            }, 25000);
        }
    }
});

export function setUpErrorHandler(app: App<unknown>): void {
    const store = useErrorHandlerStore();
    app.config.errorHandler = (err, vm, info) => {
        store.handleError({ err, info });
    };
}
