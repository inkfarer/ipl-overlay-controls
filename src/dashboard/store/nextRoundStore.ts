import { NextRound } from 'schemas';
import { SetNextRoundRequest } from 'types/messages/rounds';
import { BeginNextMatchRequest } from 'types/messages/activeRound';
import { defineStore } from 'pinia';

const nextRound = nodecg.Replicant<NextRound>('nextRound');

export const nextRoundReps = [ nextRound ];

export interface NextRoundStore {
    nextRound: NextRound;
}

export const useNextRoundStore = defineStore('nextRound', {
    state: () => ({
        nextRound: null
    } as NextRoundStore),
    actions: {
        setShowOnStream(newValue: boolean) {
            nextRound.value.showOnStream = newValue;
        },
        beginNextMatch(request: BeginNextMatchRequest): void {
            nodecg.sendMessage('beginNextMatch', request);
        },
        setNextRound(request: SetNextRoundRequest): void {
            nodecg.sendMessage('setNextRound', request);
        }
    }
});
