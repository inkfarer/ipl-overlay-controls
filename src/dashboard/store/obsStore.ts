import { NodeCGBrowser } from 'nodecg/browser';
import { ObsCredentials, ObsData } from 'schemas';
import { createStore, Store, useStore } from 'vuex';
import cloneDeep from 'lodash/cloneDeep';
import { InjectionKey } from 'vue';
import { SetObsDataRequest } from 'types/messages/obs';

const obsData = nodecg.Replicant<ObsData>('obsData');
const obsCredentials = nodecg.Replicant<ObsCredentials>('obsCredentials');

export const obsReps = [obsData, obsCredentials];

export interface ObsStore {
    obsData: ObsData
    obsCredentials: ObsCredentials
}

export const obsStore = createStore<ObsStore>({
    state: {
        obsData: null,
        obsCredentials: null
    },
    mutations: {
        setState(store, { name, val }: { name: string, val: unknown }): void {
            this.state[name] = cloneDeep(val);
        }
    },
    actions: {
        async connect(store, { address, password }: { address: string, password?: string }): Promise<void> {
            return nodecg.sendMessage('connectToObs', { address, password });
        },
        async setData(store, data: SetObsDataRequest): Promise<void> {
            return nodecg.sendMessage('setObsData', data);
        },
        async startGame(): Promise<void> {
            return nodecg.sendMessage('startGame');
        },
        async endGame(): Promise<void> {
            return nodecg.sendMessage('endGame');
        },
        setEnabled(store, enabled: boolean): void {
            nodecg.sendMessage('setObsSocketEnabled', enabled);
        }
    }
});

export const obsStoreKey: InjectionKey<Store<ObsStore>> = Symbol();

export function useObsStore(): Store<ObsStore> {
    return useStore(obsStoreKey);
}
