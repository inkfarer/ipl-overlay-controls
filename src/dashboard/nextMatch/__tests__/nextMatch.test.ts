import NextMatch from '../nextMatch.vue';
import { config, mount } from '@vue/test-utils';
import { useNextRoundStore } from '../../store/nextRoundStore';
import { createTestingPinia, TestingPinia } from '@pinia/testing';
import { IplSmallToggle } from '@iplsplatoon/vue-components';
import { TournamentDataSource } from 'types/enums/tournamentDataSource';
import difference from 'lodash/difference';
import { useHighlightedMatchStore } from '../highlightedMatchStore';

describe('NextMatch', () => {
    let pinia: TestingPinia;

    config.global.stubs = {
        HighlightedMatchPicker: true,
        ManualTeamPicker: true,
        IplErrorDisplay: true,
        IplCheckbox: true,
        IplSmallToggle: true
    };

    beforeEach(() => {
        pinia = createTestingPinia();
        config.global.plugins = [pinia];

        useNextRoundStore().$state = {
            // @ts-ignore
            nextRound: {
                showOnStream: true
            }
        };
        // @ts-ignore
        useHighlightedMatchStore().tournamentData = { meta: { source: TournamentDataSource.BATTLEFY } };
    });

    describe('template', () => {
        it('matches snapshot', () => {
            const wrapper = mount(NextMatch);

            expect(wrapper.html()).toMatchSnapshot();
        });

        it('matches snapshot when choosing teams manually', async () => {
            const wrapper = mount(NextMatch);

            wrapper.getComponent<typeof IplSmallToggle>('[data-test="choose-manually-toggle"]').vm.$emit('update:modelValue', true);
            await wrapper.vm.$nextTick();

            expect(wrapper.html()).toMatchSnapshot();
        });
    });

    const importableSources = [
        TournamentDataSource.SMASHGG,
        TournamentDataSource.BATTLEFY,
        TournamentDataSource.SENDOU_INK
    ];

    it.each(difference(Object.values(TournamentDataSource), importableSources))('disables importing from source if the source is %s', source => {
        // @ts-ignore
        useHighlightedMatchStore().tournamentData = { meta: { source } };
        const wrapper = mount(NextMatch);

        const chooseManuallyToggle = wrapper.findComponent<typeof IplSmallToggle>('[data-test="choose-manually-toggle"]');
        expect(chooseManuallyToggle.exists()).toEqual(false);
    });

    it.each(importableSources)('allows importing from source if the source is %s', source => {
        // @ts-ignore
        useHighlightedMatchStore().tournamentData = { meta: { source } };
        const wrapper = mount(NextMatch);

        const chooseManuallyToggle = wrapper.findComponent<typeof IplSmallToggle>('[data-test="choose-manually-toggle"]');
        expect(chooseManuallyToggle.exists()).toEqual(true);
    });

    it('handles changing show on stream checkbox', async () => {
        const store = useNextRoundStore();
        store.setShowOnStream = jest.fn();
        const wrapper = mount(NextMatch);

        wrapper.getComponent<typeof IplSmallToggle>('[data-test="show-on-stream-toggle"]').vm.$emit('update:modelValue', false);
        await wrapper.vm.$nextTick();

        expect(store.setShowOnStream).toHaveBeenCalledWith(false);
    });
});
