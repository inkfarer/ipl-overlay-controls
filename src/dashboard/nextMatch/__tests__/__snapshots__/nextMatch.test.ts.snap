// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NextMatch template matches snapshot 1`] = `
"<ipl-error-display-stub class="m-b-8"></ipl-error-display-stub>
<div data-v-868771ad="" class="ipl-space color-dark">
  <ipl-small-toggle-stub label="translation:chooseTeamsManuallyToggle" modelvalue="false" disabled="false" data-test="choose-manually-toggle" class="m-b-4"></ipl-small-toggle-stub>
  <ipl-small-toggle-stub label="translation:showOnStreamToggle" modelvalue="true" disabled="false" data-test="show-on-stream-toggle"></ipl-small-toggle-stub>
</div>
<highlighted-match-picker-stub class="m-t-8"></highlighted-match-picker-stub>"
`;

exports[`NextMatch template matches snapshot when choosing teams manually 1`] = `
"<ipl-error-display-stub class="m-b-8"></ipl-error-display-stub>
<div data-v-868771ad="" class="ipl-space color-dark">
  <ipl-small-toggle-stub label="translation:chooseTeamsManuallyToggle" modelvalue="true" disabled="false" data-test="choose-manually-toggle" class="m-b-4"></ipl-small-toggle-stub>
  <ipl-small-toggle-stub label="translation:showOnStreamToggle" modelvalue="true" disabled="false" data-test="show-on-stream-toggle"></ipl-small-toggle-stub>
</div>
<manual-team-picker-stub class="m-t-8"></manual-team-picker-stub>"
`;
